security/cmd/flexvolume/driver/driver.go

```diff
@@ -0,0 +1,511 @@
+}
+
+// Unmount handles the unmount command to the driver.
+func Unmount(dir string) error {
+	var emsgs []string
+	// Stop the listener.
+	// /var/lib/kubelet/pods/20154c76-bf4e-11e7-8a7e-080027631ab3/volumes/nodeagent~uds/test-volume/
+	// /var/lib/kubelet/pods/2dc75e9a-cbec-11e7-b158-0800270da466/volumes/nodeagent~uds/test-volume
+	comps := strings.Split(dir, "/")
+	if len(comps) < 6 {
+		sErr := fmt.Sprintf("Failure to notify nodeagent dir %v", dir)
+		return failure("unmount", dir, sErr)
+	}
+
+	uid := comps[5]
+	// TBD: Check if uid is the correct format.
+	naInp := &pb.WorkloadInfo{
+		Attrs:        &pb.WorkloadInfo_WorkloadAttributes{Uid: uid},
+		Workloadpath: uid,
+	}
+	if configuration.UseGrpc == true {
+		if err := sendWorkloadDeleted(naInp); err != nil {
+			sErr := "Failure to notify nodeagent: " + err.Error()
+			return failure("unmount", dir, sErr)
+		}
+	} else if err := removeCredentialFile(naInp); err != nil {
+		// Go ahead and finish the unmount; no need to hold up kubelet.
+		emsgs = append(emsgs, "Failure to delete credentials file: "+err.Error())
+	}
+
+	// unmount the bind mount
+	if err := doUnmount(filepath.Join(dir, "nodeagent")); err != nil {
+		emsgs = append(emsgs, fmt.Sprintf("unmount of %s failed", filepath.Join(dir, "nodeagent")))
+	}
+
+	// unmount the tmpfs
+	if err := doUnmount(dir); err != nil {
+		emsgs = append(emsgs, fmt.Sprintf("unmount of %s failed", dir))
+	}
+
+	// delete the directory that was created.
+	delDir := filepath.Join(configuration.NodeAgentWorkloadHomeDir, uid)
+	err := os.Remove(delDir)
+	if err != nil {
+		emsgs = append(emsgs, fmt.Sprintf("unmount del failure %s: %s", delDir, err.Error()))
+		// go head and return ok.
+	}
+
+	if len(emsgs) == 0 {
+		emsgs = append(emsgs, "Unmount Ok")
+	}
+
+	return genericSuccess("unmount", dir, strings.Join(emsgs, ","))
+}
+
+// printAndLog is used to print to stdout and to the syslog.
+func printAndLog(caller, inp, s string) {
+	fmt.Println(s)
+	logToSys(caller, inp, s)
+}
+
+// genericSuccess is to print a success response to the kubelet.
+func genericSuccess(caller, inp, msg string) error {
+	resp, err := json.Marshal(&Response{Status: "Success", Message: msg})
+	if err != nil {
+		return err
+	}
+
+	printAndLog(caller, inp, string(resp))
+	return nil
+}
+
+// failure is to print a failure response to the kubelet.
+func failure(caller, inp, msg string) error {
+	resp, err := json.Marshal(&Response{Status: "Failure", Message: msg})
+	if err != nil {
+		return errors.New(msg)
+	}
+
+	printAndLog(caller, inp, string(resp))
+	return errors.New(msg)
+}
+
+// GenericUnsupported is to print a un-supported response to the kubelet.
+func GenericUnsupported(caller, inp, msg string) error {
+	resp, err := json.Marshal(&Response{Status: "Not supported", Message: msg})
+	if err != nil {
+		return err
+	}
+
+	printAndLog(caller, inp, string(resp))
+	return nil
+}
+
+// logToSys is to write to syslog.
+func logToSys(caller, inp, opts string) {
```

Comment:
A general question: can we use golang "log" here?

Reply:
@myidpt. Thanks for the review.
The reason to use `fmt.Print` is because the interface between the kubelet and flexvolume drivers is standard output `stdout`. The kubelet expects the flexvolume driver to put json string  as output, without it being prefixed by date/time.
Hope that clarifies.

Reply:
Thanks for the clarification. That makes sense!

Reply:
@myidpt Thanks for the review comments. I've incorporated them and pushed the changes.